File: lib/vault_kit/markdown.ex
defmodule VaultKit.Markdown do
  @moduledoc """
  Extracts structured data from Obsidian-flavored markdown.

  All functions take a raw markdown string and return structured data.
  """

  @wikilink_re ~r/\[\[([^\]#|]+)(?:#([^\]|]+))?(?:\|([^\]]+))?\]\]/
  @embed_re ~r/!\[\[([^\]]+)\]\]/
  @heading_re ~r/^(\#{1,6})\s+(.+)$/m
  @checkbox_re ~r/^(\s*)- \[([ xX])\]\s+(.+)$/m
  @tag_re ~r/(?:^|(?<=\s))\#([a-zA-Z][\w\/\-]*)(?=\s|$)/m
  @frontmatter_re ~r/\A---\r?\n(.*?\r?\n)---\r?\n/s

  @doc """
  Extracts wikilinks from markdown content.

  Returns a list of maps with :name, :heading, :alias, :line, and :embed? keys.
  """
  @spec extract_wikilinks(String.t()) :: [map()]
  def extract_wikilinks(content) do
    lines = String.split(content, ~r/\r?\n/)

    embeds = extract_embeds(lines)
    links = extract_plain_links(lines)

    (embeds ++ links)
    |> Enum.sort_by(& &1.line)
  end

  defp extract_embeds(lines) do
    lines
    |> Enum.with_index(1)
    |> Enum.flat_map(fn {line, line_num} ->
      Regex.scan(@embed_re, line)
      |> Enum.map(fn [_full, target] ->
        case Regex.run(@wikilink_re, "[[#{target}]]") do
          nil -> %{name: String.trim(target), heading: nil, alias: nil}
          match -> parse_wikilink_match(match)
        end
        |> Map.merge(%{line: line_num, embed?: true})
      end)
    end)
  end

  defp extract_plain_links(lines) do
    lines
    |> Enum.with_index(1)
    |> Enum.flat_map(fn {line, line_num} ->
      cleaned = Regex.replace(@embed_re, line, "")

      Regex.scan(@wikilink_re, cleaned)
      |> Enum.map(fn match ->
        parse_wikilink_match(match)
        |> Map.merge(%{line: line_num, embed?: false})
      end)
    end)
  end

  defp parse_wikilink_match([_, name, heading, display]) do
    %{
      name: String.trim(name),
      heading: if(heading != "", do: heading, else: nil),
      alias: if(display != "", do: display, else: nil)
    }
  end

  defp parse_wikilink_match([_, name, heading]) do
    %{
      name: String.trim(name),
      heading: if(heading != "", do: heading, else: nil),
      alias: nil
    }
  end

  defp parse_wikilink_match([_, name]) do
    %{name: String.trim(name), heading: nil, alias: nil}
  end

  @doc """
  Extracts and parses YAML frontmatter from markdown content.

  Returns `{:ok, map}` if frontmatter is found and parsed, or `:none`.
  """
  @spec extract_frontmatter(String.t()) :: {:ok, map()} | :none
  def extract_frontmatter(content) do
    case Regex.run(@frontmatter_re, content) do
      [_, yaml_str] ->
        case YamlElixir.read_from_string(yaml_str) do
          {:ok, parsed} when is_map(parsed) -> {:ok, parsed}
          _ -> :none
        end

      nil ->
        :none
    end
  end

  @doc """
  Extracts headings from markdown content.

  Returns a list of maps with :level, :text, and :line keys.
  """
  @spec extract_headings(String.t()) :: [map()]
  def extract_headings(content) do
    content
    |> String.split(~r/\r?\n/)
    |> Enum.with_index(1)
    |> Enum.flat_map(fn {line, line_num} ->
      case Regex.run(@heading_re, line) do
        [_, hashes, text] ->
          [%{level: String.length(hashes), text: String.trim(text), line: line_num}]

        nil ->
          []
      end
    end)
  end

  @doc """
  Extracts checkbox/task items from markdown content.

  Returns a list of maps with :checked?, :text, :line, and :indent keys.
  """
  @spec extract_checkboxes(String.t()) :: [map()]
  def extract_checkboxes(content) do
    content
    |> String.split(~r/\r?\n/)
    |> Enum.with_index(1)
    |> Enum.flat_map(fn {line, line_num} ->
      case Regex.run(@checkbox_re, line) do
        [_, indent, check, text] ->
          [
            %{
              checked?: check in ["x", "X"],
              text: String.trim(text),
              line: line_num,
              indent: String.length(indent)
            }
          ]

        nil ->
          []
      end
    end)
  end

  @doc """
  Extracts inline tags (e.g. #tag, #nested/tag) from markdown content.

  Excludes heading markers. Returns a deduplicated list of tag strings.
  """
  @spec extract_inline_tags(String.t()) :: [String.t()]
  def extract_inline_tags(content) do
    # Strip frontmatter so YAML tags aren't matched
    body = strip_frontmatter(content)

    body
    |> String.split(~r/\r?\n/)
    |> Enum.reject(fn line -> Regex.match?(~r/^\s*\#{1,6}\s+/, line) end)
    |> Enum.flat_map(fn line ->
      Regex.scan(@tag_re, line)
      |> Enum.map(fn [_, tag] -> tag end)
    end)
    |> Enum.uniq()
  end

  @doc """
  Returns the markdown body without frontmatter.
  """
  @spec strip_frontmatter(String.t()) :: String.t()
  def strip_frontmatter(content) do
    Regex.replace(@frontmatter_re, content, "")
  end
end

File: src/markdown.ts
import type { Wikilink, Heading, Checkbox } from "./types.js";
import YAML from "yaml";

const WIKILINK_RE = /\[\[([^\]#|]+)(?:#([^\]|]+))?(?:\|([^\]]+))?\]\]/g;
const EMBED_FULL_RE = /!\[\[([^\]#|]+)(?:#([^\]|]+))?(?:\|([^\]]+))?\]\]/g;
const EMBED_RE = /!\[\[([^\]]+)\]\]/g;
const TAG_RE = /(?:^|(?<=\s))#([a-zA-Z][\w/-]*)(?=\s|$)/gm;
const FRONTMATTER_RE = /^---\r?\n([\s\S]*?\r?\n)---\r?\n/;

export function extractWikilinks(lines: string[]): Wikilink[] {
  const embeds = extractEmbeds(lines);
  const links = extractPlainLinks(lines);
  return [...embeds, ...links].sort((a, b) => a.line - b.line);
}

function extractEmbeds(lines: string[]): Wikilink[] {
  const results: Wikilink[] = [];
  for (let i = 0; i < lines.length; i++) {
    const lineNum = i + 1;
    for (const match of lines[i].matchAll(EMBED_FULL_RE)) {
      results.push({
        name: match[1].trim(),
        heading: match[2] || null,
        alias: match[3] || null,
        line: lineNum,
        embed: true,
      });
    }
  }
  return results;
}

function extractPlainLinks(lines: string[]): Wikilink[] {
  const results: Wikilink[] = [];
  for (let i = 0; i < lines.length; i++) {
    const cleaned = lines[i].replace(EMBED_RE, "");
    const lineNum = i + 1;
    for (const match of cleaned.matchAll(WIKILINK_RE)) {
      results.push({
        name: match[1].trim(),
        heading: match[2] || null,
        alias: match[3] || null,
        line: lineNum,
        embed: false,
      });
    }
  }
  return results;
}

export function extractFrontmatter(
  content: string,
): Record<string, unknown> | null {
  const match = FRONTMATTER_RE.exec(content);
  if (!match) return null;
  try {
    const parsed: unknown = YAML.parse(match[1]);
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      return parsed as Record<string, unknown>;
    }
    return null;
  } catch {
    return null;
  }
}

export function extractHeadings(lines: string[]): Heading[] {
  const results: Heading[] = [];
  const re = /^(#{1,6})\s+(.+)$/;
  for (let i = 0; i < lines.length; i++) {
    const match = re.exec(lines[i]);
    if (match) {
      results.push({
        level: match[1].length,
        text: match[2].trim(),
        line: i + 1,
      });
    }
  }
  return results;
}

export function extractCheckboxes(lines: string[]): Checkbox[] {
  const results: Checkbox[] = [];
  const re = /^(\s*)- \[([ xX])\]\s+(.+)$/;
  for (let i = 0; i < lines.length; i++) {
    const match = re.exec(lines[i]);
    if (match) {
      results.push({
        checked: match[2] === "x" || match[2] === "X",
        text: match[3].trim(),
        line: i + 1,
        indent: match[1].length,
      });
    }
  }
  return results;
}

export function extractInlineTags(lines: string[]): string[] {
  const body = stripFrontmatterLines(lines);
  const seen = new Set<string>();
  const results: string[] = [];

  for (const line of body) {
    if (/^\s*#{1,6}\s+/.test(line)) continue;
    for (const match of line.matchAll(TAG_RE)) {
      const tag = match[1];
      if (!seen.has(tag)) {
        seen.add(tag);
        results.push(tag);
      }
    }
  }
  return results;
}

export function serializeFrontmatter(fm: Record<string, unknown>): string {
  const yamlStr = YAML.stringify(fm, { lineWidth: 0 }).trimEnd();
  return `---\n${yamlStr}\n---\n`;
}

export function replaceFrontmatter(
  content: string,
  newFm: Record<string, unknown>,
): string {
  const serialized = serializeFrontmatter(newFm);
  const match = FRONTMATTER_RE.exec(content);
  if (match) {
    return serialized + content.slice(match[0].length);
  }
  return serialized + content;
}

export function stripFrontmatterLines(lines: string[]): string[] {
  if (lines.length === 0 || lines[0] !== "---") return lines;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i] === "---" || lines[i] === "---\r") {
      return lines.slice(i + 1);
    }
  }
  return lines;
}
